src/react/hooks/use-occupancy.ts

```diff
@@ -0,0 +1,109 @@
+import { Listenable } from '../listenable.js';
+import { StatusParams } from '../status-params.js';
+import { useChatConnection } from './use-chat-connection.js';
+import { useRoom } from './use-room.js';
+
+/**
+ * The options for the {@link useOccupancy} hook.
+ */
+export interface UseOccupancyParams extends StatusParams, Listenable<OccupancyListener> {
+  /**
+   * This value determines how often, in milliseconds, the occupancy metrics are updated.
+   * If not provided, the occupancy metrics will be updated on every event.
+   */
+  updateInterval?: number;
+
+  /**
+   * A listener that will be called whenever an occupancy event is received.
+   */
+  listener?: OccupancyListener;
+}
+
+/**
+ * The response type from the {@link useOccupancy} hook.
+ */
+export interface UseOccupancyResponse extends ChatStatusResponse {
+  /**
+   * Provides access to the underlying {@link Occupancy} instance of the room.
+   */
+  readonly occupancy: Occupancy;
+
+  /**
+   * Number of users connected to the room.
+   */
+  readonly connections: number;
+
+  /**
+   * Number of users present in the room.
+   */
+  readonly presenceMembers: number;
+}
+
+/**
+ * A hook that provides access to the {@link Occupancy} instance of the room.
+ * It will use the instance belonging to the nearest {@link RoomProvider} in the component tree.
+ *
+ * @param params - Allows the registering of optional callbacks and setting the optional update interval.
+ * @returns UseOccupancyResponse
+ */
+export const useOccupancy = (params?: UseOccupancyParams): UseOccupancyResponse => {
+  const { currentStatus: connectionStatus, error: connectionError } = useChatConnection({
+    onStatusChange: params?.onConnectionStatusChange,
+  });
+  const { room, roomError, roomStatus } = useRoom({
+    onStatusChange: params?.onRoomStatusChange,
+  });
+
+  const [occupancyMetrics, setOccupancyMetrics] = useState<{ connections: number; presenceMembers: number }>({
+    connections: 0,
+    presenceMembers: 0,
+  });
+
+  // if provided, subscribes the user provided discontinuity listener
+  useEffect(() => {
+    if (!params?.onDiscontinuity) return;
+    const { off } = room.occupancy.onDiscontinuity(params.onDiscontinuity);
+    return () => {
+      off();
+    };
+  }, [room, params]);
+
+  // subscribe to occupancy events, throttling the updates if an interval is provided
+  useEffect(() => {
+    const throttledEvents = throttle(setOccupancyMetrics, params?.updateInterval ?? 0);
+    const { unsubscribe } = room.occupancy.subscribe((occupancyEvent) => {
+      throttledEvents({
+        connections: occupancyEvent.connections,
+        presenceMembers: occupancyEvent.presenceMembers,
+      });
+    });
+    return () => {
+      unsubscribe();
+      throttledEvents.cancel();
+    };
+  }, [params?.updateInterval, room]);
+
+  // if provided, subscribes the user provided listener to occupancy events
+  useEffect(() => {
+    if (!params?.listener) return;
```

Comment:
Hmm, I assumed if you are registering a listener you want all events, you also have control over what happens when an event comes through easily.

Whereas, the returned metric was a shortcut way for a user to add it to some effect as a dependency, in the latter we need the param to provide a simple way to control how often it might cause a re-render.

I don't have a strong opinion on this though, so for now I will make it clear in the typedocs, but can change it after later.

Reply:
Happy for that approach - just need some clear docs that explain what it applies to :)